pytorch_lightning/trainer/progress.py

```diff
@@ -0,0 +1,192 @@
+        if self.total.started is None or self.current.started is None:
+            return
+        self.total.started += 1
+        self.current.started += 1
+
+    def increment_processed(self) -> None:
+        if self.total.processed is None or self.current.processed is None:
+            return
+        self.total.processed += 1
+        self.current.processed += 1
+
+    def increment_completed(self) -> None:
+        if self.total.completed is None or self.current.completed is None:
+            return
+        self.total.completed += 1
+        self.current.completed += 1
+
+    @classmethod
+    def from_defaults(cls, **kwargs: Optional[int]) -> 'BaseProgress':
+        return cls(total=Progress(**kwargs), current=Progress(**kwargs))
+
+
+@dataclass
+class LoopProgress:
+    """
+    Dataclass to track loop progress during execution.
+
+    These counters are local to a trainer rank. By default, they are not globally synced across all ranks.
+
+    Args:
+        epoch: Tracks epochs progress.
+        batch: Tracks batch progress.
+    """
+    epoch: BaseProgress = field(default_factory=BaseProgress)
+    batch: BaseProgress = field(default_factory=BaseProgress)
+
+    def increment_batch_ready(self) -> None:
+        self.batch.increment_ready()
+
+    def increment_batch_started(self) -> None:
+        self.batch.increment_started()
+
+    def increment_batch_processed(self) -> None:
+        self.batch.increment_processed()
+
+    def increment_batch_completed(self) -> None:
+        self.batch.increment_completed()
+
+    def increment_epoch_completed(self) -> None:
+        self.epoch.total.completed += 1
+        self.reset_on_epoch()
+
+    def reset_on_epoch(self) -> None:
+        self.batch.current.reset()
+        self.epoch.current.reset()
+
+
+@dataclass
+class OptimizationProgress:
+    """
+    Dataclass to track optimization progress.
+
+    Args:
+        optimizer: Tracks optimizer progress.
+        scheduler: Tracks scheduler progress.
+    """
+    optimizer: BaseProgress = BaseProgress.from_defaults(processed=None)
+    scheduler: BaseProgress = BaseProgress.from_defaults(started=None, processed=None)
+    zero_grad: BaseProgress = BaseProgress.from_defaults(processed=None)
+
+    @property
+    def optimizer_steps(self) -> int:
+        return self.optimizer.total.completed
+
+    @property
+    def scheduler_steps(self) -> int:
+        return self.scheduler.total.completed
+
+
+@dataclass
+class TrainProgress(BaseProgress):
+    """
+    Extends the progress with training specific attributes
+
+    Args:
+        optimization: Tracks optimization progress
+    """
+    optimization: OptimizationProgress = field(default_factory=OptimizationProgress)
+
+
+@dataclass
+class TrainLoopProgress(LoopProgress):
+    epoch: TrainProgress = field(default_factory=TrainProgress)
+
+    def reset_on_epoch(self) -> None:
+        # override to avoid resetting `epoch.current`
+        self.batch.current.reset()
```

Comment:
@ananthsub We need to override this method in the case of the training loop as `epoch.current` should track the "number of epochs this fit() call"